mitxgraders/helpers/get_number_of_args.py

```diff
@@ -0,0 +1,173 @@
+    # the __doc__ attribute of the function.
+    docstr = obj.__doc__
+    if docstr:
+        items = docstr.split('\n')
+        if items:
+            func_descr = items[0]
+            s = func_descr.replace(obj.__name__, '')
+            idx1 = s.find('(')
+            idx_default = s.find('[')
+            idx2 = s.find(')') if idx_default == -1 else idx_default
+            if idx1 != -1 and idx2 != -1 and (idx2 > idx1+1):
+                argstring = s[idx1+1:idx2]
+                # This gets the argument string
+                # Count the number of commas!
+                return argstring.count(",") + 1
+    return 0  # pragma: no cover
+
+def get_number_of_args_py2(callable_obj):
+    """
+    Get number of positional arguments of function or callable object.
+
+    NOTES:
+        - Seems to work in Python 3, but is based on inspect.getargspec which
+          raises a DeprecationWarning in Python 3.
+        - in Python 3, use the much simpler signature-based
+          get_number_of_args_py3 funciton instead.
+        - Cannot handle class constructors
+
+    Usage
+    =====
+    See documetnation for get_number_of_args.
+
+    Note, however, that this function cannot handle class constructors:
+
+    >>> class Foo(object):
+    ...     def __init__(self, x, y):
+    ...         pass
+    >>> try:                                                # doctest: +ELLIPSIS
+    ...     get_number_of_args_py2(Foo)
+    ... except ValueError as error:
+    ...     print(error)
+    Cannot detect number of arguments for <class '...Foo'>
+
+
+    """
+
+    if inspect.isbuiltin(callable_obj):
+        # Built-in function
+        func = callable_obj
+        return get_builtin_positional_args(func)
+    elif hasattr(callable_obj, "nin"):
+        # Matches RandomFunction or numpy ufunc
+        return callable_obj.nin
+    else:
+        if inspect.isfunction(callable_obj) or inspect.ismethod(callable_obj):
+            # Assume object is a function
+            func = callable_obj
+            # see https://docs.python.org/2/library/inspect.html#inspect.inspect.getargspec
+            # defaults might be None, or something weird for Mock functions
+        elif inspect.isclass(callable_obj):
+            # We don't need this anyway
+            raise ValueError("Cannot detect number of arguments for {}".format(callable_obj))
+        else:
+            # callable object instance
+            func = callable_obj.__call__
+
+        args, _, _, defaults = inspect.getargspec(func)
+
+    try:
+        num_args = len(args) - len(defaults)
+    except TypeError:
+        num_args = len(args)
+
+    # If func is a bound method, remove one argument
+    # (in Python 2.7, unbound methods have __self__ = None)
+    try:
+        if func.__self__ is not None:
+            num_args += -1
+    except AttributeError:
+        pass
+
+    return num_args
+
+def get_number_of_args_py3(callable_obj):
```

Comment:
@jolyonb BTW: This function and one line below are responsible for the drop in Python 2 coverage.

This function is probably not strictly necessary: `get_number_of_args_py2` works in Python 3 also, however it raises a `DeprecationWarning` for `getargspec`. 

Also, `get_number_of_args_py3` is about 100 times simpler, even with the annoying numpy special case.